File: 第二章_排序/选择排序练习题/SelectionSort.cpp
#include <iostream>
using namespace std; 
 
  
class SelectionSort {
public:
    int* selectionSort(int* A, int n) {
         
        for(int i=0;i<n/2;i++)
        {
            int start=i;
            int end=n-i-1;
            int max,min;
            int s,e;
 
            max=start;
            min=start;
            s=start;
            e=end;
 
            for(;s<=e;s++)
            {
                if(A[s]>A[max])
                {
                    max=s;
                }
 
                if(A[s]<A[min])
                {
                    min=s;
                }
            }
             
             
            if(max!=start&&min!=end){
                 int temp=A[max];
                A[max]=A[end];
                A[end]=temp;
                 
                temp=A[min];
                A[min]=A[start];
                A[start]=temp;
            }
            else if(max==start&&min==end){
                int temp=A[start];
                A[start]=A[end];
                A[end]=temp;
                   
            }else{
                if(max==start){
                    int temp=A[max];
                    A[max]=A[end];
                    A[end]=temp;
                 
                    temp=A[min];
                    A[min]=A[start];
                    A[start]=temp;
                }
                 
                if(min==end){
                    int temp=A[min];
                    A[min]=A[start];
                    A[start]=temp;
                     
                    temp=A[max];
                    A[max]=A[end];
                    A[end]=temp;
                }
            }
         
        }
 
        return A;
    }
};

File: 第二章_排序/选择排序练习题/SelectionSort.java
import java.util.*;
 
public class SelectionSort {
    public int[] selectionSort(int[] A, int n) {
       int temp;
       for(int i=1;i<=n/2;i++){
           int min_index=i;
           int max_index=i;
           for(int j=i-1;j<=n-i;j++){
               if(A[min_index]>A[j]){
                   min_index=j;
                    
               }
               if(A[max_index]<A[j]){
                   max_index=j;
               }
           }
           if(min_index!=i-1){
               temp=A[min_index];
               A[min_index]=A[i-1];
               A[i-1]=temp;
               if(max_index==i-1){
                   max_index=min_index;
               }
           }
           if(max_index!=n-i){
               temp=A[max_index];
               A[max_index]=A[n-i];
               A[n-i]=temp;
           }
       }
       return A;
    }
}